notebooks/platform/xarray/lib/io/pybgen_backend.py

```diff
@@ -0,0 +1,91 @@
+import dask.array as da
+import numpy as np
+from pybgen import PyBGEN
+from pybgen.parallel import ParallelPyBGEN
+import xarray as xr
+
+from .. import core
+from ..typing import PathType
+from ..compat import Requirement
+from ..dispatch import ClassBackend, register_backend
+from .core import BGEN_DOMAIN
+
+
+def _array_name(f, path):
+    return f.__qualname__ + ':' + str(path)
+
+
+class BgenReader(object):
+
+    def __init__(self, path, dtype=np.float32):
+        self.path = str(path) # pybgen needs string paths
+
+        # Use ParallelPyBGEN only to get all the variant seek positions from the BGEN index.
+        # No parallel IO happens here.
+        with ParallelPyBGEN(self.path, probs_only=False) as bgen:
+            bgen._get_all_seeks()
+            self._seeks = bgen._seeks
+            n_variants = bgen.nb_variants
+            n_samples = bgen.nb_samples
+
+            self.shape = (n_variants, n_samples)
+            self.dtype = dtype
+            self.ndim = 2
+
+            self.sample_id = bgen.samples
+            # This may need chunking for large numbers of variants
+            variants = list(bgen.iter_variant_info())
+            self.variant_id = [v.name for v in variants]
+            self.contig = [v.chrom for v in variants]
+            self.pos = [v.pos for v in variants]
+            self.a1 = [v.a1 for v in variants]
+            self.a2 = [v.a2 for v in variants]
+
+    def __getitem__(self, idx):
+        if not isinstance(idx, tuple):
+            raise IndexError(f'Indexer must be tuple (received {type(idx)})')
+        if len(idx) != self.ndim:
+            raise IndexError(f'Indexer must be two-item tuple (received {len(idx)} slices)')
+
+        # Restrict to seeks for this chunk
+        seeks_for_chunk = self._seeks[idx[0]]
+        if len(seeks_for_chunk) == 0:
+            return np.empty((0, 0), dtype=self.dtype)
+        with PyBGEN(self.path, probs_only=False) as bgen:
+            p = [probs for (_, probs) in bgen._iter_seeks(seeks_for_chunk)]
```

Comment:
I think it will be important to see what happens here with a non-diploid dataset.  We may need to reshape the result to recover the ploidy dimension or perhaps only allow diploid calls?  Do you already have a sense of what pybgen returns here in the ploidy > 2 case?

Reply:
PyBGEN only supports a ploidy of 2, see https://github.com/lemieuxl/pybgen/blob/master/pybgen/pybgen.py#L465. For non-diploid, we'll either have to extend PyBGEN, or switch to bgen-reader (but that has other issues, see above). I think we'll have a better idea of where to put our effort after we've done some scale testing with loading BGEN.

Reply:
Got it.  The only multiploid data I'm aware of would come from VCF anyhow so I think we can table worrying about that for a while.  I'm sure we'll have much bigger problems to solve scaling this out for UKB, like you mentioned.